### PR TITLE
POST endpoint for demo datasource project

### PR DESCRIPTION
**Pull request by lukesonnet**
### Features and Changes

* Adds `POST /demo-datasource-project` private endpoint, in line with https://www.notion.so/growthbook/Tech-Spec-Sample-Datasource-Creation-42936c77e753476d8be99c0b02c848eb
* Endpoint creates following assets:
  * Project with reliable project ID
  * Datasource to Postgres DB (for now my local, will update to cloud instance)
  * Metrics
  * Feature
  * Experiment
  * Experiment Snapshot
 
- Closes GB-292

TODO:
- [x] Update descriptions of experiments, identifier types, and more copy to explain the created objects
- [x] Consider adding a feature
- [x] Better error handling in endpoint
- [x] Update dates once we know the schedule of cron jobs for postgres instance

### Dependencies

- GB-300 (PR #1486)

### Testing

In-app testing for now, see Loom below.

### Screenshots

Loom: https://www.loom.com/share/5a25fcebd5ce45d3aba6a365f3fba64e

### CODE REVIEW

File: packages/back-end/src/models/ProjectModel.ts [hunk 1]
```diff
@@ -31,15 +31,21 @@ function toInterface(doc: ProjectDocument): ProjectInterface {
   return omit(ret, ["__v", "_id"]);
 }
 
+interface createProjectProps {
```
Comment (tinahollygb) [inline packages/back-end/src/models/ProjectModel.ts:34]: Naming conventions for types should use PascalCase. Can we name it `CreateProjectRequest` to align with the naming we've been doing elsewhere?

Comment (tinahollygb) [reply packages/back-end/src/models/ProjectModel.ts:34]: Sorry, disregard that last feedback re: `Request`—we are in the model.

File: packages/back-end/src/models/ProjectModel.ts [hunk 2]
```diff
@@ -31,15 +31,21 @@ function toInterface(doc: ProjectDocument): ProjectInterface {
...
 ) {
-  // TODO: sanitize fields
   const doc = await ProjectModel.create({
-    ...data,
-    organization,
-    id: uniqid("prj_"),
+    organization: organization,
+    id: data.id || uniqid("prj_"),
+    name: data.name || "",
+    description: data.description || undefined,
```
Comment (tinahollygb) [inline packages/back-end/src/models/ProjectModel.ts:48]: `|| undefined` is redundant, unless you expect `null` or `0` to be passed in, which it shouldn't if our object validation is preventing us from getting unexpected types.

I do prefer this explicit style to the previous destructuring so we know which fields we are getting 👍

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+  const { org } = getOrgFromReq(req);
+
+  try {
+    // TODO smarter management for cases when project exists
+    // TODO throw appropriate error after each step
+    await deleteProjectById(
+      getDemoDatasourceProjectIdForOrganization(org.id),
+      org.id
+    );
```
Comment (tinahollygb) [inline packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:216]: I don't think the post endpoint should be deleting the project. We can use the delete project endpoint for that.

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.router.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+const demoDatasourceProjectController = wrapController(
+  rawDemoDatasourceProjectController
+);
+
+router.post(
+  "/",
+  validateRequestMiddleware({
+    body: z
+      .object({
+        // TODO:
```
Comment (tinahollygb) [inline packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.router.ts:18]: This should be the expected fields for the project interface

File: packages/back-end/src/routers/project/project.controller.ts [hunk 5]
```diff
@@ -61,8 +61,8 @@ export const postProject = async (
   const { org } = getOrgFromReq(req);
 
   const doc = await createProject(org.id, {
-    name,
-    description,
+    name: name,
+    description: description,
```
Comment (tinahollygb) [inline packages/back-end/src/routers/project/project.controller.ts:65]: Is this change necessary?

File: packages/back-end/src/scripts/experiments.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
+[
+  {
+    "name": "checkout-layout",
+    "trackingKey": "checkout-layout",
+    "owner": "test@notarealemail.com",
```
Comment (tinahollygb) [inline packages/back-end/src/scripts/experiments.json:5]: In the event we ever start sending out emails to owners, I think it's best we don't store emails that are under domains we don't own. Might be best to do something like `test-user@growthbook.io`.

File: packages/front-end/components/GuidedGetStarted/GuidedGetStarted.tsx [hunk 7]
```diff
@@ -276,6 +276,16 @@ export default function GuidedGetStarted({
               }
             />
           )}
+          <button
+            className="my-4 btn btn-primary"
+            onClick={async () => {
+              await apiCall(`/demo-datasource-project/`, {
+                method: "POST",
+              });
```
Comment (tinahollygb) [inline packages/front-end/components/GuidedGetStarted/GuidedGetStarted.tsx:284]: It doesn't look like we're doing anything after this starts. Was this just for your testing purposes? We discussed having a separate page for this.

Comment (lukesonnet) [reply packages/front-end/components/GuidedGetStarted/GuidedGetStarted.tsx:284]: Yes this is just for testing.

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts (hunk 3 above)
Comment (lukesonnet) [reply packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:216]: So should I catch an error if we try to create the project and it already exists? Or should we just rely on us never doing that.

Comment (tinahollygb) [reply packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:216]: If our data modelling is correct, we should have a uniqueness constraint on the project ID. Have you tried to see what would happen if you try to create it? If it fails, we should have a global error handler it falls back on. I think that's the approach we're using for everything. Create endpoints should not perform delete operations.

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.router.ts (hunk 4 above)
Comment (lukesonnet) [reply packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.router.ts:18]: Hmm, I don't think it is actually (this is for the demo datasource project, not all projects). The demo datasource endpoint request should have nothing in it since we hard-code the description and just use the organization id for the project id.

File: packages/back-end/src/routers/project/project.controller.ts (hunk 5 above)
Comment (lukesonnet) [reply packages/back-end/src/routers/project/project.controller.ts:65]: nope, undone.

File: packages/back-end/src/scripts/experiments.json (hunk 6 above)
Comment (lukesonnet) [reply packages/back-end/src/scripts/experiments.json:5]: Oh, these files shouldn't have been picked up and are no longer needed in this PR, sorry.

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts (hunk 3 above)
Comment (lukesonnet) [reply packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts]: Yeah, it is handled by this error: Error: Failed to create demo datasource and project with message: E11000 duplicate key error collection: test.projects index: id_1 dup key: { id: "prj_org_2s2jlyqlbl7tcr3_demo-datasource-project" }

If the endpoint only gets exposed when the project doesn't exist, I don't mind just throwing this generic error.

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.router.ts (hunk 4 above)
Comment (tinahollygb) [reply packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.router.ts]: True, that makes sense. I think I was looking at the model, forgot we hardcoded these.

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts (hunk 3 above)
Comment (tinahollygb) [reply packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts]: That error is probably fine. It has Mongo or Mongoose error codes in it which is not ideal but it's the way we've been doing it everywhere else so I'm fine with it.

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+    environmentSettings: {
+      production: {
+        enabled: true,
+        rules: [
+          {
+            type: "force",
+            description: "",
+            id: "gbdemo-checkout-layout-employee-force-rule",
+            value: "dev",
+            condition: "is_employee = true",
```
Comment (jdorn) [inline packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:143]: Conditions must be a JSON formatted string.

```suggestion
            condition: `{"is_employee":true}`,
```

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+          projects: [project.id],
+          tags: DEMO_TAGS,
+        });
+      })
+    );
+
+    const ratioMetric = await createMetric({
+      ...DEMO_RATIO_METRIC,
+      denominator: metrics.find(
+        (x) => x.name === "Purchases - Number of Orders (72 hour window)"
```
Comment (jdorn) [inline packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:325]: This seems very fragile

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+  user: "lukesonnet",
+  host: "localhost",
+  database: "sample",
+  password: "",
+  port: 5432,
+  ssl: false,
+  defaultSchema: "",
+};
+
+const ASSET_OWNER = "datascience@growthbook.io";
```
Comment (jdorn) [inline packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:54]: I think I'd prefer just leaving the owner field as an empty string

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts (hunk 9 above)
Comment (lukesonnet) [reply packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts]: I created a constant name that can be used in both places, and made sure the denominator existed before doing anything with it.

**Review by tinahollygb — APPROVED** [review]

**Review by tinahollygb — COMMENTED** [review]
I experienced an issue when pulling this into my code. I'm guessing it won't work unless the datasource exists, and it doesn't.

image

**Review by tinahollygb — APPROVED** [review]
After the latest SSL change, it works.
